did/jwk.go

```diff
@@ -0,0 +1,179 @@
+	}
+	pubKeyJWK, err := crypto.PublicKeyToJWK(pubKey)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "converting public key to JWK")
+	}
+
+	// 2. Serialize it into a UTF-8 string
+	// 3. Encode string using base64url
+	// 4. Prepend the string with the did:jwk prefix
+	didJWK, err := CreateDIDJWK(pubKeyJWK)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "creating did:jwk")
+	}
+	return privKey, didJWK, nil
+}
+
+// CreateDIDJWK creates a did:jwk from a JWK public key by following the steps in the spec:
+// https://github.com/quartzjer/did-jwk/blob/main/spec.md
+func CreateDIDJWK(publicKeyJWK jwk.Key) (*DIDJWK, error) {
+	// 2. Serialize it into a UTF-8 string
+	pubKeyJWKBytes, err := json.Marshal(publicKeyJWK)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshalling public key JWK")
+	}
+	pubKeyJWKStr := string(pubKeyJWKBytes)
+
+	// 3. Encode string using base64url
+	encodedPubKeyJWKStr := base64.URLEncoding.EncodeToString([]byte(pubKeyJWKStr))
+
+	// 4. Prepend the string with the did:jwk prefix
+	didJWK := DIDJWK(fmt.Sprintf("%s:%s", JWKPrefix, encodedPubKeyJWKStr))
+	return &didJWK, nil
+}
+
+// Expand turns the DID JWK into a compliant DID Document
+func (d DIDJWK) Expand() (*Document, error) {
+	id := d.String()
+
+	encodedJWK, err := d.Suffix()
+	if err != nil {
+		return nil, fmt.Errorf("invalid did:jwk: %s", d)
+	}
+	decodedPubKeyJWKStr, err := base64.URLEncoding.DecodeString(encodedJWK)
+	if err != nil {
+		return nil, errors.Wrap(err, "decoding did:jwk")
+	}
+
+	var pubKeyJWK crypto.PublicKeyJWK
+	if err = json.Unmarshal(decodedPubKeyJWKStr, &pubKeyJWK); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling did:jwk")
+	}
+
+	keyReference := "#0"
+	keyID := id + keyReference
+
+	doc := Document{
+		Context: []string{KnownDIDContext, JWS2020Context},
+		ID:      id,
+		VerificationMethod: []VerificationMethod{
+			{
+				ID:           keyID,
+				Type:         cryptosuite.JSONWebKey2020Type,
+				Controller:   id,
+				PublicKeyJWK: &pubKeyJWK,
+			},
+		},
+		Authentication:       []VerificationMethodSet{keyID},
+		AssertionMethod:      []VerificationMethodSet{keyID},
+		KeyAgreement:         []VerificationMethodSet{keyID},
+		CapabilityInvocation: []VerificationMethodSet{keyID},
+		CapabilityDelegation: []VerificationMethodSet{keyID},
+	}
+
+	// If the JWK contains a use property with the value "sig" then the keyAgreement property is not included in the
+	// DID Document. If the use value is "enc" then only the keyAgreement property is included in the DID Document.
+	switch pubKeyJWK.Use {
+	case "sig":
+		doc.KeyAgreement = nil
+	case "enc":
+		doc.Authentication = nil
+		doc.AssertionMethod = nil
+		doc.CapabilityInvocation = nil
+		doc.CapabilityDelegation = nil
+	}
+
+	return &doc, nil
+}
+
+func isSupportedJWKType(kt crypto.KeyType) bool {
+	jwkTypes := GetSupportedDIDJWKTypes()
+	for _, t := range jwkTypes {
+		if t == kt {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
it happens to overlap but it's a distinct method since there's no guarantee we enable all supported key types for did key and DID JWK. for example, did:jwk can support any JWK type. did:key only supports what's in the spec.